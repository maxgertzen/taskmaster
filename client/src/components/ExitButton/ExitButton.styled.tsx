import styled from '@emotion/styled';

export const StyledExitButton = styled('div')(({ theme }) => ({
  isplay: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  gap: theme.spacing(2),
  cursor: 'pointer',
}));

export const TitleContainer = styled('div')(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1),
}));
