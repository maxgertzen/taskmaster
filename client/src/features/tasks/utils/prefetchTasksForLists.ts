import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
