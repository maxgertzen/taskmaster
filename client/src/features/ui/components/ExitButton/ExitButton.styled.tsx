import styled from '@emotion/styled';

export const StyledExitButton = styled('div')(({ theme }) => ({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  gap: theme.spacing(2),
  cursor: 'pointer',
  marginTop: 'auto',
}));

export const TitleContainer = styled('div')(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1),
}));
