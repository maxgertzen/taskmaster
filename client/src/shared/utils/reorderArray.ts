export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
