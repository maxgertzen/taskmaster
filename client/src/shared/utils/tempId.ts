export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
