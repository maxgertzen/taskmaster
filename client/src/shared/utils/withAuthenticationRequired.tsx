import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
