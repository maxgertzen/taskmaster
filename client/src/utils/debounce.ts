export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
