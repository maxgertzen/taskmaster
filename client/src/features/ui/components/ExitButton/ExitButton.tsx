import { FC } from 'react';

import { Title } from '..';
import { SpriteIcon } from '../SpriteIcon/SpriteIcon';

import { StyledExitButton, TitleContainer } from './ExitButton.styled';

interface ExitButtonProps {
  isCollapsed: boolean;
  onClick: () => void;
}

export const ExitButton: FC<ExitButtonProps> = ({ isCollapsed, onClick }) => {
  return (
    <StyledExitButton role='button' onClick={onClick}>
      <TitleContainer>
        <SpriteIcon name='exit' />
        {!isCollapsed && <Title variant='h6'>Exit</Title>}
      </TitleContainer>
    </StyledExitButton>
  );
};
