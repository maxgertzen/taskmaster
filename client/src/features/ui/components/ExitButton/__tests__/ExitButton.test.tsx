import { describe, it, expect, vi } from 'vitest';

import { render, screen, within } from '@/tests/utils';

import { ExitButton } from '../ExitButton';

describe('ExitButton', () => {
  describe.each([
    ['expanded', false],
    ['collapsed', true],
  ])('when %s (isCollapsed: %s)', (_, isCollapsed) => {
    it.each([
      ['light', undefined],
      ['dark', 'dark'],
    ])('renders correctly in %s theme', (_, theme) => {
      const { container } = render(<ExitButton isCollapsed={isCollapsed} />, {
        theme: theme as 'light' | 'dark',
      });
      expect(container).toMatchSnapshot();
    });

    it('renders icon correctly', () => {
      render(<ExitButton isCollapsed={isCollapsed} />);
      const button = screen.getByTestId('exit-button');
      expect(within(button).getByTestId('exit-icon')).toBeInTheDocument();
    });

    it('handles text visibility correctly', () => {
      render(<ExitButton isCollapsed={isCollapsed} />);
      const button = screen.getByTestId('exit-button');

      if (isCollapsed) {
        expect(within(button).queryByText('Exit')).not.toBeInTheDocument();
      } else {
        expect(within(button).getByText('Exit')).toBeInTheDocument();
      }
    });

    it('calls onClick when clicked', async () => {
      const handleClick = vi.fn();
      const { user } = render(
        <ExitButton isCollapsed={isCollapsed} onClick={handleClick} />
      );

      await user.click(screen.getByTestId('exit-button'));
      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });
});
