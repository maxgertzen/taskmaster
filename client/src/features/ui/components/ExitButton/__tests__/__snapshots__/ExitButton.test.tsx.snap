// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExitButton > when collapsed (isCollapsed: true) > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-xee8n6"
    data-testid="exit-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-exit"
        role="button"
        size="3"
      >
        <svg
          data-testid="test-svg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ExitButton > when collapsed (isCollapsed: true) > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-xee8n6"
    data-testid="exit-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-exit"
        role="button"
        size="3"
      >
        <svg
          data-testid="test-svg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ExitButton > when expanded (isCollapsed: false) > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-xee8n6"
    data-testid="exit-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-exit"
        role="button"
        size="3"
      >
        <svg
          data-testid="test-svg"
        />
      </div>
      <h6
        css="[object Object]"
      >
        Exit
      </h6>
    </div>
  </div>
</div>
`;

exports[`ExitButton > when expanded (isCollapsed: false) > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-xee8n6"
    data-testid="exit-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-exit"
        role="button"
        size="3"
      >
        <svg
          data-testid="test-svg"
        />
      </div>
      <h6
        css="[object Object]"
      >
        Exit
      </h6>
    </div>
  </div>
</div>
`;
